In [1]:
import numpy as np
import pandas as pd

In [2]:
TRAIN_DATASET_PATH = 'realestatepriceprediction/train.csv'
TEST_DATASET_PATH = 'realestatepriceprediction/test.csv'
PREPARED_TRAIN_DATASET_PATH = 'realestatepriceprediction/prepared_train.csv'

In [3]:
# Создадим датафрейм с тренировочными данными
df = pd.read_csv(TRAIN_DATASET_PATH)
df.head()

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
0  14038          35    2.0  47.981561   29.442751            6.0      7   
1  15053          41    3.0  65.683640   40.049543            8.0      7   
2   4765          53    2.0  44.947953   29.197612            0.0      8   
3   5809          58    2.0  53.352981   52.731512            9.0      8   
4  10783          99    1.0  39.649192   23.776169            7.0     11   

   HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  \
0         9.0       1969   0.089040         B         B        33      7976   
1         9.0       1978   0.000070         B         B        46     10309   
2        12.0       1968   0.049637         B         B        34      7759   
3        17.0       1977   0.437885         B         B        23      5735   
4        12.0       1976   0.012339         B         B        35      5776   

   Social_3  Healthcare_1  Helthcare_2  Shops_1 Shops_2          Price  
0         5           NaN            0       11       B  184966.930730  
1         1         240.0            1       16       B  300009.450063  
2         0         229.0            1        3       B  220925.908524  
3         3        1084.0            0        5       B  175616.227217  
4         1        2078.0            2        4       B  150226.531644

In [4]:
# Удалим признак id
df = df.drop('Id', axis=1)
df.head()

DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  HouseFloor  \
0          35    2.0  47.981561   29.442751            6.0      7         9.0   
1          41    3.0  65.683640   40.049543            8.0      7         9.0   
2          53    2.0  44.947953   29.197612            0.0      8        12.0   
3          58    2.0  53.352981   52.731512            9.0      8        17.0   
4          99    1.0  39.649192   23.776169            7.0     11        12.0   

   HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
0       1969   0.089040         B         B        33      7976         5   
1       1978   0.000070         B         B        46     10309         1   
2       1968   0.049637         B         B        34      7759         0   
3       1977   0.437885         B         B        23      5735         3   
4       1976   0.012339         B         B        35      5776         1   

   Healthcare_1  Helthcare_2  Shops_1 Shops_2          Price  
0           NaN            0       11       B  184966.930730  
1         240.0            1       16       B  300009.450063  
2         229.0            1        3       B  220925.908524  
3        1084.0            0        5       B  175616.227217  
4        2078.0            2        4       B  150226.531644

In [5]:
# Проверим полноту данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
DistrictId       10000 non-null int64
Rooms            10000 non-null float64
Square           10000 non-null float64
LifeSquare       7887 non-null float64
KitchenSquare    10000 non-null float64
Floor            10000 non-null int64
HouseFloor       10000 non-null float64
HouseYear        10000 non-null int64
Ecology_1        10000 non-null float64
Ecology_2        10000 non-null object
Ecology_3        10000 non-null object
Social_1         10000 non-null int64
Social_2         10000 non-null int64
Social_3         10000 non-null int64
Healthcare_1     5202 non-null float64
Helthcare_2      10000 non-null int64
Shops_1          10000 non-null int64
Shops_2          10000 non-null object
Price            10000 non-null float64
dtypes: float64(8), int64(8), object(3)
memory usage: 1.4+ MB


In [6]:
# В параметре Healthcare_1 отсутствует почти половина данных. Данный параметр не может быть оценен объективно.
# Удалим его
df = df.drop('Healthcare_1', axis=1)

In [7]:
df.describe().T

count           mean            std           min  \
DistrictId     10000.0      50.400800      43.587592      0.000000   
Rooms          10000.0       1.890500       0.839512      0.000000   
Square         10000.0      56.315775      21.058732      1.136859   
LifeSquare      7887.0      37.199645      86.241209      0.370619   
KitchenSquare  10000.0       6.273300      28.560917      0.000000   
Floor          10000.0       8.526700       5.241148      1.000000   
HouseFloor     10000.0      12.609400       6.775974      0.000000   
HouseYear      10000.0    3990.166300  200500.261427   1910.000000   
Ecology_1      10000.0       0.118858       0.119025      0.000000   
Social_1       10000.0      24.687000      17.532614      0.000000   
Social_2       10000.0    5352.157400    4006.799803    168.000000   
Social_3       10000.0       8.039200      23.831875      0.000000   
Helthcare_2    10000.0       1.319500       1.493601      0.000000   
Shops_1        10000.0       4.231300       4.806341      0.000000   
Price          10000.0  214138.857399   92872.293865  59174.778028   

                         25%            50%            75%           max  
DistrictId         20.000000      36.000000      75.000000  2.090000e+02  
Rooms               1.000000       2.000000       2.000000  1.900000e+01  
Square             41.774881      52.513310      65.900625  6.410652e+02  
LifeSquare         22.769832      32.781260      45.128803  7.480592e+03  
KitchenSquare       1.000000       6.000000       9.000000  2.014000e+03  
Floor               4.000000       7.000000      12.000000  4.200000e+01  
HouseFloor          9.000000      13.000000      17.000000  1.170000e+02  
HouseYear        1974.000000    1977.000000    2001.000000  2.005201e+07  
Ecology_1           0.017647       0.075424       0.195781  5.218671e-01  
Social_1            6.000000      25.000000      36.000000  7.400000e+01  
Social_2         1564.000000    5285.000000    7227.000000  1.908300e+04  
Social_3            0.000000       2.000000       5.000000  1.410000e+02  
Helthcare_2         0.000000       1.000000       2.000000  6.000000e+00  
Shops_1             1.000000       3.000000       6.000000  2.300000e+01  
Price          153872.633942  192269.644879  249135.462171  6.332335e+05

У нас не полные данные LifeSquare. Не хватает примерно трети. Это значительное количество. Проверим достоверность данных, чтобы принять решение - удалить данные или дополнить. Для этого найдем недостоверные данные:  
1. LifeSquare > Square  
2. LifeSquare > max Square
3. LifeSquare < 10

In [8]:
df.loc[(df['LifeSquare'] > df['Square']) | (df['LifeSquare'] < 10) | (df['LifeSquare'] > 641)]

DistrictId  Rooms      Square  LifeSquare  KitchenSquare  Floor  \
26            27    3.0  106.958871    0.641822            0.0     17   
28            23    1.0   42.530043   43.967759            1.0      3   
44            38    3.0  104.211396  106.340403            0.0     20   
52             1    2.0   61.400054   65.224603            0.0     17   
95            34    2.0   61.334968    4.091945            1.0     26   
...          ...    ...         ...         ...            ...    ...   
9889          23    2.0   51.440463   53.134243           51.0      3   
9895          27    3.0  123.430072  125.806981          123.0      5   
9902          73    1.0   44.098768   44.267551            1.0      7   
9910          27    4.0  200.334539  201.627361           25.0      1   
9954          27    3.0   84.250710    4.187885            1.0     15   

      HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  \
26           0.0       2018   0.072158         B         B         2   
28           9.0       2014   0.034656         B         B         0   
44           0.0       2017   0.060753         B         B        15   
52          22.0       2016   0.007122         B         B         1   
95           1.0       2014   0.069753         B         B        53   
...          ...        ...        ...       ...       ...       ...   
9889        17.0       2017   0.005767         B         B         1   
9895        10.0       2015   0.017647         B         B         2   
9902        24.0       2014   0.042032         B         B        37   
9910         2.0       2013   0.041116         B         B        53   
9954         1.0       2016   0.072158         B         B         2   

      Social_2  Social_3  Helthcare_2  Shops_1 Shops_2          Price  
26         629         1            0        0       A  337299.867936  
28         168         0            0        0       B   95338.198549  
44        2787         2            0        7       B  435462.048070  
52         264         0            0        1       B  199215.452229  
95       13670         4            1       11       B  286564.441972  
...        ...       ...          ...      ...     ...            ...  
9889       388         0            0        0       B   88150.012510  
9895       469         0            0        0       B  234194.837047  
9902      6856        84            2        5       B  381937.404161  
9910     14892         4            1        4       B  528560.506016  
9954       629         1            0        0       A  236647.781363  

[680 rows x 18 columns]

Еще 680 данных недостоверны. Когда треть данных отсутствует или недостоверна лучше их не использовать. Удалим LifeSquare.

In [2]:
df = df.drop('LifeSquare', axis=1)
df.info()

NameError: name 'df' is not defined

#### Обработаем выбросы

In [10]:
df.describe().T

count           mean            std           min  \
DistrictId     10000.0      50.400800      43.587592      0.000000   
Rooms          10000.0       1.890500       0.839512      0.000000   
Square         10000.0      56.315775      21.058732      1.136859   
KitchenSquare  10000.0       6.273300      28.560917      0.000000   
Floor          10000.0       8.526700       5.241148      1.000000   
HouseFloor     10000.0      12.609400       6.775974      0.000000   
HouseYear      10000.0    3990.166300  200500.261427   1910.000000   
Ecology_1      10000.0       0.118858       0.119025      0.000000   
Social_1       10000.0      24.687000      17.532614      0.000000   
Social_2       10000.0    5352.157400    4006.799803    168.000000   
Social_3       10000.0       8.039200      23.831875      0.000000   
Helthcare_2    10000.0       1.319500       1.493601      0.000000   
Shops_1        10000.0       4.231300       4.806341      0.000000   
Price          10000.0  214138.857399   92872.293865  59174.778028   

                         25%            50%            75%           max  
DistrictId         20.000000      36.000000      75.000000  2.090000e+02  
Rooms               1.000000       2.000000       2.000000  1.900000e+01  
Square             41.774881      52.513310      65.900625  6.410652e+02  
KitchenSquare       1.000000       6.000000       9.000000  2.014000e+03  
Floor               4.000000       7.000000      12.000000  4.200000e+01  
HouseFloor          9.000000      13.000000      17.000000  1.170000e+02  
HouseYear        1974.000000    1977.000000    2001.000000  2.005201e+07  
Ecology_1           0.017647       0.075424       0.195781  5.218671e-01  
Social_1            6.000000      25.000000      36.000000  7.400000e+01  
Social_2         1564.000000    5285.000000    7227.000000  1.908300e+04  
Social_3            0.000000       2.000000       5.000000  1.410000e+02  
Helthcare_2         0.000000       1.000000       2.000000  6.000000e+00  
Shops_1             1.000000       3.000000       6.000000  2.300000e+01  
Price          153872.633942  192269.644879  249135.462171  6.332335e+05

DistrictId может принимать любые значения, кроме отрицательных. Таких нет

In [11]:
# Рассмотрим квартиры с Square меньше 10
df.loc[df['Square']  < 10]

DistrictId  Rooms    Square  KitchenSquare  Floor  HouseFloor  \
212           88    2.0  5.497061            1.0     24        22.0   
1316          27    1.0  4.633498            1.0     18         1.0   
1608           6    1.0  2.596351            1.0      3        25.0   
3280          27    1.0  4.380726            1.0     10        17.0   
3413           5    1.0  5.129222            1.0      1         1.0   
4853          27    0.0  2.377248            0.0      1         0.0   
4900          27    3.0  4.390331            1.0      8        19.0   
6392           1    1.0  1.136859            1.0      3         1.0   
8030           1    3.0  4.823679            0.0      6        17.0   
8283          34    1.0  1.988943            1.0     21         4.0   
9294          45    1.0  2.954309            1.0      3         1.0   
9613          38    1.0  4.502784            1.0     28         1.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
212        1977   0.127376         B         B        43      8429         3   
1316       1977   0.011654         B         B         4       915         0   
1608       2014   0.243205         B         B         5      1564         0   
3280       2013   0.211401         B         B         9      1892         0   
3413       1977   0.150818         B         B        16      3433         4   
4853       1977   0.017647         B         B         2       469         0   
4900       2016   0.211401         B         B         9      1892         0   
6392       1977   0.007122         B         B         1       264         0   
8030       1977   0.007122         B         B         1       264         0   
8283       1977   0.069753         B         B        53     13670         4   
9294       1977   0.195781         B         B        23      5212         6   
9613       1977   0.060753         B         B        15      2787         2   

      Helthcare_2  Shops_1 Shops_2          Price  
212             3        9       B  412511.088764  
1316            0        0       B  107604.269441  
1608            0        0       B  137597.601458  
3280            0        1       B   97560.720383  
3413            4        5       B  369472.403061  
4853            0        0       B  126596.941798  
4900            0        1       B  161379.067034  
6392            0        1       B  181434.825589  
8030            0        1       B  237716.681261  
8283            1       11       B  458378.777006  
9294            3        2       B  438005.182323  
9613            0        7       B  483283.488083

In [12]:
# Нет ни одной Square пригодной для проживания. 
# Предположим, что в данных ошибка в порядке и умножим эти данные на 10.
df.loc[df['Square']  < 10, 'Square'] = df.loc[df['Square']  < 10] * 10

In [13]:
# Проверим
df.loc[df['Square']  < 20]

DistrictId  Rooms     Square  KitchenSquare  Floor  HouseFloor  \
1891           0    1.0  17.498542            0.0      7         5.0   
3166         150    1.0  17.046188            8.0      7        12.0   
3794         101    1.0  18.509177            3.0      5         5.0   
4739          81    3.0  13.784865            7.0      4         5.0   
5617          57    3.0  16.547242            8.0      5         5.0   
6392           1    1.0  11.368588            1.0      3         1.0   
6945          62    2.0  17.414181            1.0      4        20.0   
7767          21    1.0  16.117154            0.0      2         5.0   
7897          21    1.0  18.983962            2.0      1         5.0   
8283          34    1.0  19.889427            1.0     21         4.0   
9516           0    1.0  19.927423            0.0      2         5.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
1891       1961   0.319809         B         B        25      4756        16   
3166       1982   0.300323         B         B        52     10311         6   
3794       1970   0.225825         A         B        41      6558         7   
4739       1960   0.319809         B         B        25      4756        16   
5617       1962   0.133215         B         B        49     11395         3   
6392       1977   0.007122         B         B         1       264         0   
6945       1977   0.072158         B         B         2       629         1   
7767       1965   0.194489         B         B        47      8004         3   
7897       1967   0.194489         B         B        47      8004         3   
8283       1977   0.069753         B         B        53     13670         4   
9516       1962   0.319809         B         B        25      4756        16   

      Helthcare_2  Shops_1 Shops_2          Price  
1891            5        8       B   88871.750230  
3166            1        9       B   74222.110465  
3794            3        2       B  115845.861527  
4739            5        8       B   78388.806186  
5617            3        4       A   77835.185808  
6392            0        1       B  181434.825589  
6945            0        0       A  180750.471749  
7767            3        5       B   79913.267856  
7897            3        5       B  112411.221700  
8283            1       11       B  458378.777006  
9516            5        8       B  102395.788856

In [14]:
# Найдем квартиры с площадью комнаты меньше 10
df.loc[((df['Square'] - df['KitchenSquare']) / df['Rooms']) < 10]

DistrictId  Rooms      Square  KitchenSquare  Floor  HouseFloor  \
227            2    3.0   79.722243           72.0     12        16.0   
377           57   10.0   59.056975           10.0     22        22.0   
551            6    1.0   48.128114           48.0     21         1.0   
1064          62    1.0   47.100719         2014.0      4         1.0   
1077          27    1.0   40.081042           37.0     18        19.0   
1369          27    2.0   68.841073           66.0      4         2.0   
1454           1   19.0   42.006046            7.0     17        17.0   
1455          54    2.0   79.810535           78.0     10        15.0   
1860         161    2.0   53.216778           53.0      7        17.0   
2170          99    6.0   59.414334            6.0      7         9.0   
2247         109    3.0   32.795221            6.0      2         9.0   
2916          72    3.0   97.490674           96.0     22        25.0   
2969          62    3.0  114.734473          112.0      3         3.0   
3166         150    1.0   17.046188            8.0      7        12.0   
3271          27    5.0   42.003293            0.0      5        12.0   
3379          27    1.0   36.138409           36.0     19        19.0   
4079          23    2.0   67.146049           63.0      5        17.0   
4110          27    2.0   66.787523           60.0     14        20.0   
4142          88    1.0   42.357487           41.0     10        19.0   
4431          53    5.0   39.906082            1.0      5         1.0   
4588          48    1.0   41.186904           41.0      2         1.0   
4651          58    3.0  116.405693          112.0      3         3.0   
4739          81    3.0   13.784865            7.0      4         5.0   
5149          42    1.0   38.071692           73.0      9        10.0   
5162          27    2.0   61.070298           58.0      8        17.0   
5480          10    1.0   35.813564           35.0      6        14.0   
5617          57    3.0   16.547242            8.0      5         5.0   
5816          27    1.0   43.292516           40.0     12        11.0   
5849          46    1.0   42.983283           37.0     15        16.0   
5888          52    2.0   64.859242           62.0     15        17.0   
5904          27    1.0   54.418214           51.0     10        17.0   
6036          34    2.0   60.603363           58.0     14        22.0   
6240          96    3.0   78.445675           75.0      9        12.0   
6945          62    2.0   17.414181            1.0      4        20.0   
7088          27    1.0   38.220258           84.0      4        17.0   
7480          62    1.0   46.013187           43.0      4        17.0   
8148           1    3.0   31.113950            6.0      6         5.0   
8232          34    2.0   56.738764           54.0     26        25.0   
8551          96    4.0   44.128947           10.0     10        13.0   
8584          81    1.0   32.276663         1970.0      6         1.0   
8849           9   10.0   60.871266           10.0      3         2.0   
9111          48    1.0   42.069359           39.0      1        10.0   
9889          23    2.0   51.440463           51.0      3        17.0   
9895          27    3.0  123.430072          123.0      5        10.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
227        1987   0.130618         B         B        39     10418         9   
377        2002   0.090799         B         B        74     19083         2   
551        2013   0.243205         B         B         5      1564         0   
1064       2014   0.072158         B         B         2       629         1   
1077       2019   0.211401         B         B         9      1892         0   
1369       2014   0.017647         B         B         2       469         0   
1454       2014   0.007122         B         B         1       264         0   
1455       2014   0.006076         B         B        30      5285         0   


In [15]:
# Изменим количество комнат на 1
df.loc[((df['Square'] - df['KitchenSquare']) / df['Rooms']) < 10, 'Rooms'] = 1

In [16]:
df.loc[((df['Square'] - df['KitchenSquare']) / df['Rooms']) < 10]

DistrictId  Rooms      Square  KitchenSquare  Floor  HouseFloor  \
227            2    1.0   79.722243           72.0     12        16.0   
551            6    1.0   48.128114           48.0     21         1.0   
1064          62    1.0   47.100719         2014.0      4         1.0   
1077          27    1.0   40.081042           37.0     18        19.0   
1369          27    1.0   68.841073           66.0      4         2.0   
1455          54    1.0   79.810535           78.0     10        15.0   
1860         161    1.0   53.216778           53.0      7        17.0   
2916          72    1.0   97.490674           96.0     22        25.0   
2969          62    1.0  114.734473          112.0      3         3.0   
3166         150    1.0   17.046188            8.0      7        12.0   
3379          27    1.0   36.138409           36.0     19        19.0   
4079          23    1.0   67.146049           63.0      5        17.0   
4110          27    1.0   66.787523           60.0     14        20.0   
4142          88    1.0   42.357487           41.0     10        19.0   
4588          48    1.0   41.186904           41.0      2         1.0   
4651          58    1.0  116.405693          112.0      3         3.0   
4739          81    1.0   13.784865            7.0      4         5.0   
5149          42    1.0   38.071692           73.0      9        10.0   
5162          27    1.0   61.070298           58.0      8        17.0   
5480          10    1.0   35.813564           35.0      6        14.0   
5617          57    1.0   16.547242            8.0      5         5.0   
5816          27    1.0   43.292516           40.0     12        11.0   
5849          46    1.0   42.983283           37.0     15        16.0   
5888          52    1.0   64.859242           62.0     15        17.0   
5904          27    1.0   54.418214           51.0     10        17.0   
6036          34    1.0   60.603363           58.0     14        22.0   
6240          96    1.0   78.445675           75.0      9        12.0   
7088          27    1.0   38.220258           84.0      4        17.0   
7480          62    1.0   46.013187           43.0      4        17.0   
8232          34    1.0   56.738764           54.0     26        25.0   
8584          81    1.0   32.276663         1970.0      6         1.0   
9111          48    1.0   42.069359           39.0      1        10.0   
9889          23    1.0   51.440463           51.0      3        17.0   
9895          27    1.0  123.430072          123.0      5        10.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
227        1987   0.130618         B         B        39     10418         9   
551        2013   0.243205         B         B         5      1564         0   
1064       2014   0.072158         B         B         2       629         1   
1077       2019   0.211401         B         B         9      1892         0   
1369       2014   0.017647         B         B         2       469         0   
1455       2014   0.006076         B         B        30      5285         0   
1860       1994   0.000699         B         B        14      3369        24   
2916       2019   0.210473         B         B        11      2398         2   
2969       2015   0.072158         B         B         2       629         1   
3166       1982   0.300323         B         B        52     10311         6   
3379       2017   0.072158         B         B         2       629         1   
4079       2019   0.034656         B         B         0       168         0   
4110       2015   0.017647         B         B         2       469         0   
4142       2016   0.127376         B         B        43      8429         3   
4588       1977   0.041125         B         B        46      9515         5   
4651       2016   0.437885         B         B        23      5735         3   
4739       1960   0.319809         B         B        25      4756        16   
5149      

In [17]:
# Можно допустить, что комната одна и она совмещена с кухней, поэтому их площади равны или близки по значениюю
# Выберем квартиры, где кухня больше площади.
df.loc[df['Square'] < df['KitchenSquare']]

DistrictId  Rooms     Square  KitchenSquare  Floor  HouseFloor  \
1064          62    1.0  47.100719         2014.0      4         1.0   
5149          42    1.0  38.071692           73.0      9        10.0   
7088          27    1.0  38.220258           84.0      4        17.0   
8584          81    1.0  32.276663         1970.0      6         1.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
1064       2014   0.072158         B         B         2       629         1   
5149       2006   0.158249         B         B        21      5731         0   
7088       2018   0.011654         B         B         4       915         0   
8584       1977   0.006076         B         B        30      5285         0   

      Helthcare_2  Shops_1 Shops_2          Price  
1064            0        0       A  108337.484207  
5149            1        0       B  160488.033165  
7088            0        0       B   99079.960518  
8584            6        6       B  105539.556275

In [18]:
# В двух площади кухни явно перепутаны с годами. Заменим их на 0.
# В двух других скорее всего значения площади и кухни перепутаны местами
# Исправим это
df.loc[df['KitchenSquare'] > 1000, 'KitchenSquare'] = 0

In [19]:
df.loc[df['Square'] < df['KitchenSquare']]

DistrictId  Rooms     Square  KitchenSquare  Floor  HouseFloor  \
5149          42    1.0  38.071692           73.0      9        10.0   
7088          27    1.0  38.220258           84.0      4        17.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
5149       2006   0.158249         B         B        21      5731         0   
7088       2018   0.011654         B         B         4       915         0   

      Helthcare_2  Shops_1 Shops_2          Price  
5149            1        0       B  160488.033165  
7088            0        0       B   99079.960518

In [20]:
a = np.array([df.loc[df['Square'] < df['KitchenSquare'], 'Square']])
a

array([[38.07169235, 38.22025827]])

In [21]:
df.loc[df['Square'] < df['KitchenSquare'], 'Square'] = df['KitchenSquare']
df.loc[df['Square'] == df['KitchenSquare']]

DistrictId  Rooms  Square  KitchenSquare  Floor  HouseFloor  HouseYear  \
5149          42    1.0    73.0           73.0      9        10.0       2006   
7088          27    1.0    84.0           84.0      4        17.0       2018   

      Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
5149   0.158249         B         B        21      5731         0   
7088   0.011654         B         B         4       915         0   

      Helthcare_2  Shops_1 Shops_2          Price  
5149            1        0       B  160488.033165  
7088            0        0       B   99079.960518

In [22]:
b, c = a[0, 0], a[0, 1]

In [23]:
df.loc[df['Square'] == df['KitchenSquare'], 'KitchenSquare'] = b, c

In [24]:
# Проверим
df.loc[(df['Square'] == 73) | (df['Square'] == 84)]

DistrictId  Rooms  Square  KitchenSquare  Floor  HouseFloor  HouseYear  \
5149          42    1.0    73.0      38.071692      9        10.0       2006   
7088          27    1.0    84.0      38.220258      4        17.0       2018   

      Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
5149   0.158249         B         B        21      5731         0   
7088   0.011654         B         B         4       915         0   

      Helthcare_2  Shops_1 Shops_2          Price  
5149            1        0       B  160488.033165  
7088            0        0       B   99079.960518

In [25]:
df.describe().T

count           mean            std           min  \
DistrictId     10000.0      50.400800      43.587592      0.000000   
Rooms          10000.0       1.882200       0.811905      0.000000   
Square         10000.0      56.363816      20.996411     11.368588   
KitchenSquare  10000.0       5.866829       5.091614      0.000000   
Floor          10000.0       8.526700       5.241148      1.000000   
HouseFloor     10000.0      12.609400       6.775974      0.000000   
HouseYear      10000.0    3990.166300  200500.261427   1910.000000   
Ecology_1      10000.0       0.118858       0.119025      0.000000   
Social_1       10000.0      24.687000      17.532614      0.000000   
Social_2       10000.0    5352.157400    4006.799803    168.000000   
Social_3       10000.0       8.039200      23.831875      0.000000   
Helthcare_2    10000.0       1.319500       1.493601      0.000000   
Shops_1        10000.0       4.231300       4.806341      0.000000   
Price          10000.0  214138.857399   92872.293865  59174.778028   

                         25%            50%            75%           max  
DistrictId         20.000000      36.000000      75.000000  2.090000e+02  
Rooms               1.000000       2.000000       2.000000  5.000000e+00  
Square             41.787090      52.535436      65.912854  6.410652e+02  
KitchenSquare       1.000000       6.000000       9.000000  1.230000e+02  
Floor               4.000000       7.000000      12.000000  4.200000e+01  
HouseFloor          9.000000      13.000000      17.000000  1.170000e+02  
HouseYear        1974.000000    1977.000000    2001.000000  2.005201e+07  
Ecology_1           0.017647       0.075424       0.195781  5.218671e-01  
Social_1            6.000000      25.000000      36.000000  7.400000e+01  
Social_2         1564.000000    5285.000000    7227.000000  1.908300e+04  
Social_3            0.000000       2.000000       5.000000  1.410000e+02  
Helthcare_2         0.000000       1.000000       2.000000  6.000000e+00  
Shops_1             1.000000       3.000000       6.000000  2.300000e+01  
Price          153872.633942  192269.644879  249135.462171  6.332335e+05

In [26]:
# Проверим дома с 0 этажей и этажем квартиры выше, чем этажей в доме
df.loc[(df['HouseFloor'] == 0) | (df['HouseFloor'] < df['Floor'])]

DistrictId  Rooms      Square  KitchenSquare  Floor  HouseFloor  \
7             74    2.0   80.312926            0.0     14         0.0   
14            27    1.0   53.769097            1.0      5         4.0   
16            27    2.0   49.360648            5.0      6         5.0   
21            27    2.0   64.711835            1.0     15         1.0   
23            54    3.0  118.907612            0.0      2         0.0   
...          ...    ...         ...            ...    ...         ...   
9984          52    3.0   64.074977            6.0     11         9.0   
9985          27    1.0   38.968338            1.0     10         1.0   
9989          96    3.0   90.055233           11.0     26        24.0   
9991           3    3.0  101.622794           12.0     18        17.0   
9995          32    2.0   50.401785            5.0      6         5.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
7          1977   0.075779         B         B         6      1437         3   
14         1977   0.072158         B         B         2       629         1   
16         1983   0.051815         B         B         5      1227         0   
21         1977   0.211401         B         B         9      1892         0   
23         1977   0.006076         B         B        30      5285         0   
...         ...        ...       ...       ...       ...       ...       ...   
9984       1983   0.371149         B         B        34      7065         1   
9985       1977   0.211401         B         B         9      1892         0   
9989       2005   0.041116         B         B        53     14892         4   
9991       2000   0.265089         B         B        37      5288         0   
9995       1968   0.135650         B         B        46      7960         6   

      Helthcare_2  Shops_1 Shops_2          Price  
7               0        2       B  221244.156664  
14              0        0       A  140647.565937  
16              0        0       B  117000.381287  
21              0        1       B  127200.026511  
23              6        6       B  571069.052600  
...           ...      ...     ...            ...  
9984            2        5       B  226434.880156  
9985            0        1       B   95529.569405  
9989            1        4       B  441108.911237  
9991            3        2       B  212883.501504  
9995            3       11       B  196684.316040  

[1825 rows x 17 columns]

In [27]:
# Заменим значения HouseFloor на Floor
df.loc[(df['HouseFloor'] == 0) | (df['HouseFloor'] < df['Floor']), 'HouseFloor'] = df['Floor']

In [28]:
df.describe().T

count           mean            std           min  \
DistrictId     10000.0      50.400800      43.587592      0.000000   
Rooms          10000.0       1.882200       0.811905      0.000000   
Square         10000.0      56.363816      20.996411     11.368588   
KitchenSquare  10000.0       5.866829       5.091614      0.000000   
Floor          10000.0       8.526700       5.241148      1.000000   
HouseFloor     10000.0      13.435700       6.094012      1.000000   
HouseYear      10000.0    3990.166300  200500.261427   1910.000000   
Ecology_1      10000.0       0.118858       0.119025      0.000000   
Social_1       10000.0      24.687000      17.532614      0.000000   
Social_2       10000.0    5352.157400    4006.799803    168.000000   
Social_3       10000.0       8.039200      23.831875      0.000000   
Helthcare_2    10000.0       1.319500       1.493601      0.000000   
Shops_1        10000.0       4.231300       4.806341      0.000000   
Price          10000.0  214138.857399   92872.293865  59174.778028   

                         25%            50%            75%           max  
DistrictId         20.000000      36.000000      75.000000  2.090000e+02  
Rooms               1.000000       2.000000       2.000000  5.000000e+00  
Square             41.787090      52.535436      65.912854  6.410652e+02  
KitchenSquare       1.000000       6.000000       9.000000  1.230000e+02  
Floor               4.000000       7.000000      12.000000  4.200000e+01  
HouseFloor          9.000000      14.000000      17.000000  1.170000e+02  
HouseYear        1974.000000    1977.000000    2001.000000  2.005201e+07  
Ecology_1           0.017647       0.075424       0.195781  5.218671e-01  
Social_1            6.000000      25.000000      36.000000  7.400000e+01  
Social_2         1564.000000    5285.000000    7227.000000  1.908300e+04  
Social_3            0.000000       2.000000       5.000000  1.410000e+02  
Helthcare_2         0.000000       1.000000       2.000000  6.000000e+00  
Shops_1             1.000000       3.000000       6.000000  2.300000e+01  
Price          153872.633942  192269.644879  249135.462171  6.332335e+05

In [29]:
# Найдем выбросы в HouseYear
df.loc[df['HouseYear'] > 2019]

DistrictId  Rooms     Square  KitchenSquare  Floor  HouseFloor  \
585           53    2.0  59.597891           10.0      3        24.0   
707           62    1.0  49.275155            1.0      5         5.0   
790          169    3.0  87.371123           12.0      4        24.0   
918           53    2.0  64.233072           13.0     13        24.0   
1497         109    1.0  37.265070            9.0      9        12.0   
1499         120    2.0  85.087643           10.0      7        10.0   
2493          53    1.0  35.672788            8.0      4        24.0   
2877           1    3.0  83.260743            1.0     15        21.0   
3357          62    1.0  55.911215           10.0      6        17.0   
3485          62    1.0  44.019287            1.0      7        12.0   
3859          59    1.0  48.327498            1.0      3        10.0   
4053          45    2.0  82.880909            1.0     19        22.0   
4189         147    2.0  44.791836            5.0      4         9.0   
4376          53    3.0  77.214218           12.0      6        24.0   
4493          53    2.0  60.363889           12.0     14        24.0   
5701          62    1.0  49.298503            9.0      9        12.0   
6735          74    1.0  42.799882           13.0      3         3.0   
7076          53    2.0  57.815313            8.0      6        24.0   
7459          53    1.0  38.004250            8.0      4        24.0   
8253          23    3.0  74.222388           13.0      5         5.0   
9674          53    2.0  57.126663           10.0     11        24.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
585        2020   0.049637         B         B        34      7759         0   
707        2020   0.072158         B         B         2       629         1   
790        2020   0.178917         B         B        18      3594         4   
918        2020   0.049637         B         B        34      7759         0   
1497   20052011   0.136330         B         B        30      6141        10   
1499       2020   0.341072         B         B        27      5664        48   
2493       2020   0.049637         B         B        34      7759         0   
2877       2020   0.007122         B         B         1       264         0   
3357       2020   0.072158         B         B         2       629         1   
3485       2020   0.072158         B         B         2       629         1   
3859       2020   0.019509         B         B        37      7687        11   
4053       2020   0.195781         B         B        23      5212         6   
4189       4968   0.319809         B         B        25      4756        16   
4376       2020   0.049637         B         B        34      7759         0   
4493       2020   0.049637         B         B        34      7759         0   
5701       2020   0.072158         B         B         2       629         1   
6735       2020   0.075779         B         B         6      1437         3   
7076       2020   0.049637         B         B        34      7759         0   
7459       2020   0.049637         B         B        34      7759         0   
8253       2020   0.005767         B         B         1       388         0   
9674       2020   0.049637         B         B        34      7759         0   

      Helthcare_2  Shops_1 Shops_2          Price  
585             1        3       B  243114.222804  
707             0        0       A  135353.885620  
790             1        5       B  295304.011974  
918             1        3       B  283334.496716  
1497            3        6       B  254084.534396  
1499            3       11       B  536020.260344  
2493            1        3       B  170847.416766  
2877            0        1       B  242243.090839  
3357            0        0       A  138770.715048  
3485            0        0       A  135588.215289  
3859            5        5       B  227402.467599  
4053            3        2       B  

In [30]:
# Заменим нереальные даты
df.loc[df['HouseYear'] == 20052011, 'HouseYear'] = 2005
df.loc[df['HouseYear'] == 4968, 'HouseYear'] = 1968
df.loc[df['HouseYear'] > 2019]

DistrictId  Rooms     Square  KitchenSquare  Floor  HouseFloor  \
585           53    2.0  59.597891           10.0      3        24.0   
707           62    1.0  49.275155            1.0      5         5.0   
790          169    3.0  87.371123           12.0      4        24.0   
918           53    2.0  64.233072           13.0     13        24.0   
1499         120    2.0  85.087643           10.0      7        10.0   
2493          53    1.0  35.672788            8.0      4        24.0   
2877           1    3.0  83.260743            1.0     15        21.0   
3357          62    1.0  55.911215           10.0      6        17.0   
3485          62    1.0  44.019287            1.0      7        12.0   
3859          59    1.0  48.327498            1.0      3        10.0   
4053          45    2.0  82.880909            1.0     19        22.0   
4376          53    3.0  77.214218           12.0      6        24.0   
4493          53    2.0  60.363889           12.0     14        24.0   
5701          62    1.0  49.298503            9.0      9        12.0   
6735          74    1.0  42.799882           13.0      3         3.0   
7076          53    2.0  57.815313            8.0      6        24.0   
7459          53    1.0  38.004250            8.0      4        24.0   
8253          23    3.0  74.222388           13.0      5         5.0   
9674          53    2.0  57.126663           10.0     11        24.0   

      HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  Social_3  \
585        2020   0.049637         B         B        34      7759         0   
707        2020   0.072158         B         B         2       629         1   
790        2020   0.178917         B         B        18      3594         4   
918        2020   0.049637         B         B        34      7759         0   
1499       2020   0.341072         B         B        27      5664        48   
2493       2020   0.049637         B         B        34      7759         0   
2877       2020   0.007122         B         B         1       264         0   
3357       2020   0.072158         B         B         2       629         1   
3485       2020   0.072158         B         B         2       629         1   
3859       2020   0.019509         B         B        37      7687        11   
4053       2020   0.195781         B         B        23      5212         6   
4376       2020   0.049637         B         B        34      7759         0   
4493       2020   0.049637         B         B        34      7759         0   
5701       2020   0.072158         B         B         2       629         1   
6735       2020   0.075779         B         B         6      1437         3   
7076       2020   0.049637         B         B        34      7759         0   
7459       2020   0.049637         B         B        34      7759         0   
8253       2020   0.005767         B         B         1       388         0   
9674       2020   0.049637         B         B        34      7759         0   

      Helthcare_2  Shops_1 Shops_2          Price  
585             1        3       B  243114.222804  
707             0        0       A  135353.885620  
790             1        5       B  295304.011974  
918             1        3       B  283334.496716  
1499            3       11       B  536020.260344  
2493            1        3       B  170847.416766  
2877            0        1       B  242243.090839  
3357            0        0       A  138770.715048  
3485            0        0       A  135588.215289  
3859            5        5       B  227402.467599  
4053            3        2       B  441850.464440  
4376            1        3       B  338025.692848  
4493            1        3       B  266202.142371  
5701            0        0       A  151779.182589  
6735            0        2       B  162260.682278  
7076            1        3       B  251249.365740  
7459            1        3       B  167829.123917  
8253            0        0  

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
DistrictId       10000 non-null int64
Rooms            10000 non-null float64
Square           10000 non-null float64
KitchenSquare    10000 non-null float64
Floor            10000 non-null int64
HouseFloor       10000 non-null float64
HouseYear        10000 non-null int64
Ecology_1        10000 non-null float64
Ecology_2        10000 non-null object
Ecology_3        10000 non-null object
Social_1         10000 non-null int64
Social_2         10000 non-null int64
Social_3         10000 non-null int64
Helthcare_2      10000 non-null int64
Shops_1          10000 non-null int64
Shops_2          10000 non-null object
Price            10000 non-null float64
dtypes: float64(6), int64(8), object(3)
memory usage: 1.3+ MB


In [32]:
# Рассмотрим признаки типа object
df['Ecology_2'].value_counts()

B    9903
A      97
Name: Ecology_2, dtype: int64

In [33]:
df['Ecology_3'].value_counts()

B    9725
A     275
Name: Ecology_3, dtype: int64

In [34]:
df['Shops_2'].value_counts()

B    9175
A     825
Name: Shops_2, dtype: int64

In [35]:
# Заменим В на 1 и А на 0 и тип на int
df['Ecology_2'] = (df['Ecology_2'] == 'B').astype(int)

In [36]:
df['Ecology_2'].value_counts()

1    9903
0      97
Name: Ecology_2, dtype: int64

In [37]:
df['Ecology_3'] = (df['Ecology_3'] == 'B').astype(int)

In [38]:
df['Ecology_3'].value_counts()

1    9725
0     275
Name: Ecology_3, dtype: int64

In [39]:
df['Shops_2'] = (df['Shops_2'] == 'B').astype(int)

In [40]:
df['Shops_2'].value_counts()

1    9175
0     825
Name: Shops_2, dtype: int64

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
DistrictId       10000 non-null int64
Rooms            10000 non-null float64
Square           10000 non-null float64
KitchenSquare    10000 non-null float64
Floor            10000 non-null int64
HouseFloor       10000 non-null float64
HouseYear        10000 non-null int64
Ecology_1        10000 non-null float64
Ecology_2        10000 non-null int64
Ecology_3        10000 non-null int64
Social_1         10000 non-null int64
Social_2         10000 non-null int64
Social_3         10000 non-null int64
Helthcare_2      10000 non-null int64
Shops_1          10000 non-null int64
Shops_2          10000 non-null int64
Price            10000 non-null float64
dtypes: float64(6), int64(11)
memory usage: 1.3 MB


In [42]:
df.describe().T

count           mean           std           min  \
DistrictId     10000.0      50.400800     43.587592      0.000000   
Rooms          10000.0       1.882200      0.811905      0.000000   
Square         10000.0      56.363816     20.996411     11.368588   
KitchenSquare  10000.0       5.866829      5.091614      0.000000   
Floor          10000.0       8.526700      5.241148      1.000000   
HouseFloor     10000.0      13.435700      6.094012      1.000000   
HouseYear      10000.0    1984.865700     18.411517   1910.000000   
Ecology_1      10000.0       0.118858      0.119025      0.000000   
Ecology_2      10000.0       0.990300      0.098015      0.000000   
Ecology_3      10000.0       0.972500      0.163543      0.000000   
Social_1       10000.0      24.687000     17.532614      0.000000   
Social_2       10000.0    5352.157400   4006.799803    168.000000   
Social_3       10000.0       8.039200     23.831875      0.000000   
Helthcare_2    10000.0       1.319500      1.493601      0.000000   
Shops_1        10000.0       4.231300      4.806341      0.000000   
Shops_2        10000.0       0.917500      0.275139      0.000000   
Price          10000.0  214138.857399  92872.293865  59174.778028   

                         25%            50%            75%            max  
DistrictId         20.000000      36.000000      75.000000     209.000000  
Rooms               1.000000       2.000000       2.000000       5.000000  
Square             41.787090      52.535436      65.912854     641.065193  
KitchenSquare       1.000000       6.000000       9.000000     123.000000  
Floor               4.000000       7.000000      12.000000      42.000000  
HouseFloor          9.000000      14.000000      17.000000     117.000000  
HouseYear        1974.000000    1977.000000    2001.000000    2020.000000  
Ecology_1           0.017647       0.075424       0.195781       0.521867  
Ecology_2           1.000000       1.000000       1.000000       1.000000  
Ecology_3           1.000000       1.000000       1.000000       1.000000  
Social_1            6.000000      25.000000      36.000000      74.000000  
Social_2         1564.000000    5285.000000    7227.000000   19083.000000  
Social_3            0.000000       2.000000       5.000000     141.000000  
Helthcare_2         0.000000       1.000000       2.000000       6.000000  
Shops_1             1.000000       3.000000       6.000000      23.000000  
Shops_2             1.000000       1.000000       1.000000       1.000000  
Price          153872.633942  192269.644879  249135.462171  633233.466570

In [43]:
# Сохраним данные
df.to_csv(PREPARED_TRAIN_DATASET_PATH, index=False, encoding='utf-8')